gateway/api/management/commands/update_jobs_statuses.py

```diff
@@ -36,15 +38,35 @@ def handle(self, *args, **options):
                 else:
                     success = False
 
-                if not success:
-                    kill_ray_cluster(job.compute_resource.title)
-                    job.compute_resource.delete()
-                    job.compute_resource = None
-                    job_status = Job.FAILED
-                    job.logs = (
-                        f"{job.logs}\nSomething went wrong during updating job status."
+                timeout = config.PROGRAM_TIMEOUT
+                if job.updated:
+                    endtime = job.updated + timedelta(days=timeout)
+                else:
+                    endtime = job.created + timedelta(days=timeout)
+                now = datetime.now(tz=endtime.tzinfo)
+                if endtime < now:
+                    success = False
+                    logger.warning(
+                        "Job [%s] time out: last update was at [%s], timeout length: [%s] days",
+                        job.id,
+                        job.updated,
+                        timeout,
                     )
 
+                if not success:
+                    if config.RAY_CLUSTER_NO_DELETE_ON_COMPLETE:
+                        logger.debug(
+                            "RAY_CLUSTER_NO_DELETE_ON_COMPLETE is enabled, "
+                            + "so cluster [%s] will not be removed",
+                            job.compute_resource.title,
+                        )
+                    else:
+                        kill_ray_cluster(job.compute_resource.title)
```

Comment:
Can we instead of killing cluster put job to `STOPPED` status and append log to job with something like `Job reached maximum runtime and stopped.`? 

Reason for this is one cluster can share multiple jobs, so we can incidentally kill other jobs too :) 

For killing clusters we have `free_resources` command, which will handle resource deallocation (killing clusters) where there is not actevely running jobs on it.  
https://github.com/Qiskit-Extensions/quantum-serverless/blob/913964586c65cbafba8b543cb3fb1b9135350fc1/gateway/api/management/commands/free_resources.py#L38

Reply:
Oh, I see.  I didn't looked at the meaning of "success = False" correctly.